nx/lib/nx.ex

```diff
@@ -16464,4 +16464,123 @@ defmodule Nx do
       end)
     end
   end
+
+  @doc """
+  Returns the logarithm of the sum of the exponentials of tensor elements.
+
+  If the `:axes` option is given, it aggregates over
+  the given dimensions, effectively removing them.
+  `axes: [0]` implies aggregating over the highest order
+  dimension and so forth. If the axis is negative, then
+  counts the axis from the back. For example, `axes: [-1]`
+  will always aggregate all rows.
+
+  You may optionally set `:keep_axes` to true, which will
+  retain the rank of the input tensor by setting the reduced
+  axes to size 1.
+
+  ## Examples
+
+      iex> Nx.logsumexp(Nx.tensor([1, 2, 3, 4, 5, 6]))
+      #Nx.Tensor<
+        f32
+        6.456193447113037
+      >
+
+      iex> Nx.logsumexp(Nx.tensor([1, 2, 3, 4, 5, 6]), exp_scaling_factor: 0.5)
+      #Nx.Tensor<
+        f32
+        5.7630462646484375
+      >
+
+      iex> t = Nx.tensor([1, 2, 3, 4, 5, 6])
+      iex> a = Nx.tensor([-1, -1, -1, 1, 1, 1])
+      iex> Nx.logsumexp(t, exp_scaling_factor: a)
+      #Nx.Tensor<
+        f32
+        6.356536865234375
+      >
+
+      iex> Nx.logsumexp(Nx.tensor([[1, 2], [3, 4], [5, 6]]))
+      #Nx.Tensor<
+        f32
+        6.456193447113037
+      >
+
+  ### Aggregating over an axis
+
+      iex> t = Nx.tensor([[1, 2], [3, 4], [5, 6]], names: [:x, :y])
+      iex> Nx.logsumexp(t, axes: [:x])
+      #Nx.Tensor<
+        f32[y: 2]
+        [5.1429314613342285, 6.1429314613342285]
+      >
+
+      iex> t = Nx.tensor([[1, 2], [3, 4], [5, 6]], names: [:x, :y])
+      iex> Nx.logsumexp(t, axes: [:y])
+      #Nx.Tensor<
+        f32[x: 3]
+        [2.3132617473602295, 4.31326150894165, 6.31326150894165]
+      >
+
+      iex> t = Nx.tensor([[[1, 2], [3, 4]], [[5, 6], [7, 8]]], names: [:x, :y, :z])
+      iex> Nx.logsumexp(t, axes: [:x, :z])
+      #Nx.Tensor<
+        f32[y: 2]
+        [6.331411361694336, 8.331411361694336]
+      >
+
+  ### Keeping axes
+
+      iex> t = Nx.tensor([[[1, 2], [3, 4]], [[5, 6], [7, 8]]], names: [:x, :y, :z])
+      iex> Nx.logsumexp(t, axes: [:x, :z], keep_axes: true)
+      #Nx.Tensor<
+        f32[x: 1][y: 2][z: 1]
+        [
+          [
+            [6.331411361694336],
+            [8.331411361694336]
+          ]
+        ]
+      >
+
+  ### Vectorized tensors
+
+      iex> t = Nx.vectorize(Nx.tensor([[1, 2], [3, 4], [5, 6]]), :x)
+      iex> Nx.logsumexp(t, axes: [0], keep_axes: true)
+      #Nx.Tensor<
+        vectorized[x: 3]
+        f32[1]
+        [
+          [2.3132617473602295],
+          [4.31326150894165],
+          [6.31326150894165]
+        ]
+      >
+  """
+  @doc type: :aggregation
+  def logsumexp(tensor, opts \\ []) do
+    type = type(tensor)
+    opts = keyword!(opts, [:axes, :exp_scaling_factor, :keep_axes])
+    axes = opts[:axes]
+    keep_axes = opts[:keep_axes]
+    max = reduce_max(tensor, axes: axes, keep_axes: true)
+    mask = is_infinity(max)
+    max = select(mask, Nx.tensor(0, type: type), max)
```

Comment:
I have a question regarding this treatment of infinities.
Does it make sense to actually treat them here, given that the function is conceptually undefined for positive infinities and just becomes log(0) for negative infinities?

Reply:
What I meant is the following. If there is infinity in the tensor, max will be infinity as well. To keep it numerically stable, max is subtracted from the elements, and subtracting infinity from infinity gives NaN. I suppose the intended thing to return would be infinity as you stated.

Reply:
> Also, it would be nice to add a unit-test that's not a doctest for this case specifically, showing what happens if one entry and if all entries on the reduction axis are infinity and/or neg_infinity

Yes, indeed. Will write unit test for that as well as one checking if axes kept are being squeezed.

Reply:
Let's test this with the following tensor, aggregating on axes: [1]. You can just hardcode the result in the actual test.

vals = [:infinity, :neg_infinity, 0]
Nx.tensor(for x <- vals, y <- vals, z <- vals, do: [x, y, z])